storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java

```diff
@@ -0,0 +1,117 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.log.internals;
+
+import org.apache.kafka.common.KafkaException;
+
+import java.util.Objects;
+
+/*
+ * A log offset structure, including:
+ *  1. the message offset
+ *  2. the base message offset of the located segment
+ *  3. the physical position on the located segment
+ */
+public class LogOffsetMetadata {
+
+    //TODO KAFKA-14484 remove once UnifiedLog has been moved to the storage module
+    private static final long UNIFIED_LOG_UNKNOWN_OFFSET = -1L;
+
+    public static final LogOffsetMetadata UNKNOWN_OFFSET_METADATA = new LogOffsetMetadata(-1L, 0L, 0);
+
+    private static final int UNKNOWN_FILE_POSITION = -1;
+
+    private final long messageOffset;
+    private final long segmentBaseOffset;
+    private final int relativePositionInSegment;
+
+    public LogOffsetMetadata(long messageOffset) {
+        this(messageOffset, UNIFIED_LOG_UNKNOWN_OFFSET, UNKNOWN_FILE_POSITION);
+    }
+
+    public LogOffsetMetadata(long messageOffset,
+                             long segmentBaseOffset,
+                             int relativePositionInSegment) {
+        this.messageOffset = messageOffset;
+        this.segmentBaseOffset = segmentBaseOffset;
+        this.relativePositionInSegment = relativePositionInSegment;
+    }
+
+    public long messageOffset() {
+        return messageOffset;
+    }
+
+    public long segmentBaseOffset() {
+        return segmentBaseOffset;
+    }
+
+    public int relativePositionInSegment() {
+        return relativePositionInSegment;
+    }
+
+    // check if this offset is already on an older segment compared with the given offset
+    public boolean onOlderSegment(LogOffsetMetadata that) {
+        if (messageOffsetOnly())
+            throw new KafkaException(this + " cannot compare its segment info with " + that + " since it only has message offset info");
+
+        return this.segmentBaseOffset < that.segmentBaseOffset;
+    }
+
+    // check if this offset is on the same segment with the given offset
+    private boolean onSameSegment(LogOffsetMetadata that) {
+        if (messageOffsetOnly())
+            throw new KafkaException(this + " cannot compare its segment info with " + that + " since it only has message offset info");
+
+        return this.segmentBaseOffset == that.segmentBaseOffset;
+    }
+
+    // compute the number of bytes between this offset to the given offset
+    // if they are on the same segment and this offset precedes the given offset
+    public int positionDiff(LogOffsetMetadata that) {
+        if (!onSameSegment(that))
+            throw new KafkaException(this + " cannot compare its segment position with " + that + " since they are not on the same segment");
+
```

Comment:
Here you miss the `if(messageOffsetOnly)` check.

Reply:
Let's check if any test failed. If it did not, let's add coverage, please.

Reply:
The `onSameSegment()` method already calls `messageOffsetOnly()` so this condition should be true if we reach this code.

Reply:
That said the message won't be the same and it will perhaps be a bit confusing? That is, perhaps we should do the message offset check first.

Reply:
The `if(messageOffsetOnly)` check here was effectively unreachable code. If this was true, `onSameSegment()` would have thrown already (with the exact same message).

Since `onSameSegment()` was not marked private, I guess it made sense to have the check there too, "just in case". But this method is only called from `positionDiff()`, so I've marked it private now and thought we did not need to have the redundant/unreachable check.

Reply:
I understood that part. My point was that the original goal was to have a specific error message for this method even if that didn't work out due to the other method throwing first. If you do the message offset only check first, then the problem is solved.